subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c

```diff
@@ -146,140 +146,132 @@ gst_msdkdec_get_oldest_frame (GstVideoDecoder * decoder)
+  mfxFrameSurface1 *mfx_surface = NULL;
+#endif
+  gint n = 0;
+  guint retry_times = gst_util_uint64_scale_ceil (GST_USECOND,
+      thiz->param.mfx.FrameInfo.FrameRateExtD,
+      thiz->param.mfx.FrameInfo.FrameRateExtN);
+
+  /* Free un-unsed msdk surfaces firstly, hence the associated mfx
+   * surfaces will be moved from used list to available list */
+  if (!gst_msdkdec_free_unlocked_msdk_surfaces (thiz, FALSE)) {
+    while (n < retry_times) {
+      /* It is MediaSDK/oneVPL's requirement that only the pre-allocated
+       * surfaces can be used during the whole decoding process.
+       * In the case of decoder plus multi-encoders, it is possible
+       * that all surfaces are used by downstreams and no more surfaces
+       * available for decoder. So here we need to wait until there is at
+       * least one surface is free for decoder.
+       */
+      n++;
+      g_usleep (1000);
+      if (gst_msdkdec_free_unlocked_msdk_surfaces (thiz, TRUE))
+        break;
+    }
+  }
+#ifndef _WIN32
+  if ((gst_buffer_pool_acquire_buffer (thiz->alloc_pool, &out_buffer, 0))
+      != GST_FLOW_OK) {
+    GST_ERROR_OBJECT (thiz, "Failed to allocate output buffer");
+    return NULL;
   }
 
-  if (!frame->output_buffer) {
-    /* Free un-unsed msdk surfaces firstly, hence the associated mfx
-     * surfaces will be moved from used list to available list */
-    gst_msdkdec_free_unlocked_msdk_surfaces (thiz);
+  mem = gst_buffer_peek_memory (out_buffer, 0);
+  msdk_surface = g_slice_new0 (GstMsdkSurface);
+  msdk_surface->from_qdata = FALSE;
 
-    flow = gst_video_decoder_allocate_output_frame (decoder, frame);
-    if (flow != GST_FLOW_OK) {
-      gst_video_codec_frame_unref (frame);
-      return flow;
-    }
+  if ((mfx_surface = gst_mini_object_get_qdata (GST_MINI_OBJECT_CAST (mem),
+              GST_MSDK_FRAME_SURFACE))) {
+    msdk_surface->surface = mfx_surface;
+    msdk_surface->from_qdata = TRUE;
+  } else {
+    GST_ERROR ("No available surfaces");
+    g_slice_free (GstMsdkSurface, msdk_surface);
+    return NULL;
+  }
+#else
+  if (!gst_buffer_pool_is_active (thiz->pool) &&
+      !gst_buffer_pool_set_active (thiz->pool, TRUE)) {
+    GST_ERROR_OBJECT (thiz, "Failed to activate buffer pool");
+    return NULL;
   }
 
-  *buffer = gst_buffer_ref (frame->output_buffer);
-  gst_buffer_replace (&frame->output_buffer, NULL);
-  gst_video_codec_frame_unref (frame);
+  if ((gst_buffer_pool_acquire_buffer (thiz->pool, &out_buffer, 0))
+      != GST_FLOW_OK) {
+    GST_ERROR_OBJECT (thiz, "Failed to allocate output buffer");
+    return NULL;
+  }
 
-  return GST_FLOW_OK;
-}
+  GstVideoCodecState *output_state =
+      gst_video_decoder_get_output_state (GST_VIDEO_DECODER (thiz));
 
-static MsdkSurface *
-get_surface (GstMsdkDec * thiz, GstBuffer * buffer)
-{
-  MsdkSurface *i;
-  GstVideoCodecState *output_state = NULL;
-  gboolean success;
+  msdk_surface =
+      gst_msdk_import_sys_mem_to_msdk_surface (out_buffer, output_state->info);
```

Comment:
This is not right. The gst_msdk_import_sys_mem_to_msdk_surface returns a malloced msdk_surface. And you are trying to give the returned structure to another malloced pointer.